### PR TITLE
initial commit - Azure test (app/security group)

### PR DESCRIPTION
**Pull request by rguthriemsft**
Initial commit of azure support and test cases for app and security group.  Pull request https://github.com/google/citest/pull/257 is required for this to work properly.

@EricJizbaMSFT PTAL @ewiseblatt PTAL

### CODE REVIEW

File: testing/citest/spinnaker_testing/azure_scenario_support.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        '--azure_location', 
+        dest='spinnaker_azure_location',
+        help='The azure location to be used.')
+    builder.add_argument(
+        '--test_azure_rg_location', 
+        default = defaults.get('TEST_AZURE_RG_LOCATION', 'westus'),
+        help='The location of the azure resource group where test resources should be created.')
+    # builder.add_argument(
+    #     '--spinnaker_azure_account_name', 
+    #     dest='spinnaker_azure_account_name',
```
Comment (EricJizbaMSFT) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: Uncomment this arg and change to `default=defaults.get('SPINNAKER_AZURE_ACCOUNT', None),`

File: testing/citest/spinnaker_testing/azure_scenario_support.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      defaults: [dict] Default binding value overrides.
+         This is used to initialize the default commandline parameters.
+    """
+    #
+    # Operation Parameters
+    #
+
+    # pylint: disable=line-too-long
+    # builder.add_argument(
+    #     '--azure_subscription_name',
```
Comment (EricJizbaMSFT) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: nit: remove unnecessary comments

File: testing/citest/tests/azure_smoke_test.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        'Security Group Created', retryable_for_secs=30)
+     .collect_resources(
+         az_resource='network',
+         command='nsg',
+         args=['show', '--name', self.TEST_SECURITY_GROUP, '--resource-group', self.TEST_SECURITY_GROUP_RG],
+         no_resources_ok=True)
+     .contains_pred_list([
+         jp.DICT_MATCHES({
+             'name': jp.STR_SUBSTR(self.TEST_SECURITY_GROUP)#,
+             #'securityRules': jp.STR_SUBSTR("protocol='tcp'")
```
Comment (ewiseblatt) [inline testing/citest/tests/azure_smoke_test.py]: I'm not sure what you were trying to do here.
Ideally your API returns back a json document so you have a dictionary with name value pairs (e.g. priority: 100) rather than a string (priority=100). If you have one big string you can either add a bunch of individual regexes into it or ideally have your agent be able to convert the string into a dictionary. I just checked in a python implementation of aws agent rather than CLI which is pretty simple. It's more simplistic than the gcp one. If your CLI cant give you json back then you might consider that route. It should be straightforward now that you've written the CLI one.

I think the multi regex solution here would look like
.contains_path_pred('securityRules',
    jp.AND(jp.STR_SUBSTR("protocol='tcp'"),
                 jp.STR_SUBSTR("priority=100"),
                 jp.STR_SUBSTR("destinationPortRange='80=80'")))

The reason a MATCH would be better, besides being simpler to read, is that the reporting is more straight forward, especially in the event of a failure.

File: testing/citest/README.md [hunk 4]
```diff
@@ -60,6 +60,7 @@ you may need additional tools.
 Platform | Tools | Installation Command
 ---------|-------|---------------------
 Amazon Web Services | awscli | ```sudo apt-get install -y awscli```
+Microsoft Azure | az | [See instructions](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli)
```
Comment (rguthriemsft) [inline testing/citest/README.md]: done

File: testing/citest/README.md [hunk 5]
```diff
@@ -203,6 +207,7 @@ spinnaker_google_credentials | The name of the Spinnaker [clouddriver] account t
 spinnaker_kubernetes_credentials |  The name of the Spinnaker [clouddriver] account that you wish to use for Kubernetes operations. If not specified, this will use the configured primary account.
 spinnaker_aws_credentials |  The name of the Spinnaker [clouddriver] account that you wish to use for Amazon Web Services operations. If not specified, this will use the configured primary account.
 spinnaker_os_account | The name of the Spinnaker [clouddriver] account that you wish to use for OpenStack operations. If not specified, this will use the configured primary account.
+spinnaker_azure_account | The name of the Spinnaker [clouddriver] account that you wish to use for the Azure operations. If not specified, this will use the configures primary account.
```
Comment (rguthriemsft) [inline testing/citest/README.md]: done

File: testing/citest/spinnaker_testing/azure_scenario_support.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        dest='spinnaker_azure_tenant_id',
+        help='The AAD tenant id.')
+    builder.add_argument(
+        '--azure_client_id', 
+        dest='spinnaker_azure_client_id',
+        help='The service principal id.')
+    builder.add_argument(
+        '--azure_app_key', 
+        dest='spinnaker_azure_app_key',
+        help='The service principal secret.')
```
Comment (rguthriemsft) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: No we don't support file method.  We could do environment variables but the way we plan to run these test is part of a jenkins job that deploys a templated spinnaker install and then scp's the -local.yaml file and grabs values and passes them into python/test case execution.

File: testing/citest/spinnaker_testing/azure_scenario_support.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    builder.add_argument(
+        '--azure_location', 
+        dest='spinnaker_azure_location',
+        help='The azure location to be used.')
+    builder.add_argument(
+        '--test_azure_rg_location', 
+        default = defaults.get('TEST_AZURE_RG_LOCATION', 'westus'),
+        help='The location of the azure resource group where test resources should be created.')
+    # builder.add_argument(
```
Comment (rguthriemsft) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: done

File: testing/citest/tests/azure_smoke_test.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Smoke test to see if Spinnaker can interoperate with Microsoft Azure.
+
+----- Documentation to update --------
```
Comment (rguthriemsft) [inline testing/citest/tests/azure_smoke_test.py]: added documentation

File: testing/citest/README.md (hunk 5 above)
Comment (rguthriemsft) [inline testing/citest/README.md]: done

File: testing/citest/spinnaker_testing/azure_scenario_support.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    # pylint: disable=line-too-long
+    # builder.add_argument(
+    #     '--azure_subscription_name', 
+    #     dest='spinnaker_azure_account',
+    #     help='The name of your subscriptoin.')
+    builder.add_argument(
+        '--azure_subscription_id', 
+        dest='spinnaker_azure_subscription_id',
+        help='The subscription id of your subscriptoin.')
```
Comment (rguthriemsft) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: done

File: testing/citest/tests/azure_smoke_test.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+      agent: [GateAgent] The agent for invoking the test operations on Gate.
+    """
+    super(AzureSmokeTestScenario, self).__init__(bindings, agent)
+    bindings = self.bindings
+
+    # We'll call out the app name because it is widely used
+    # because it scopes the context of our activities.
+    # pylint: disable=invalid-name
+    self.TEST_APP = bindings['TEST_APP']+AzureSmokeTestScenario.DEFAULT_TEST_ID
+    self.TEST_STACK = bindings['TEST_STACK']+AzureSmokeTestScenario.DEFAULT_TEST_ID
```
Comment (rguthriemsft) [inline testing/citest/tests/azure_smoke_test.py]: fixed

File: testing/citest/tests/azure_smoke_test.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    }]
+    builder = az.AzContractBuilder(self.az_observer)
+    (builder.new_clause_builder(
+        'Security Group Created', retryable_for_secs=30)
+     .collect_resources(
+         az_resource='network',
+         command='nsg',
+         args=['show', '--name', self.TEST_SECURITY_GROUP, '--resource-group', self.TEST_SECURITY_GROUP_RG],
+         no_resources_ok=True)
+     .contains_pred_list([
```
Comment (rguthriemsft) [inline testing/citest/tests/azure_smoke_test.py]: we plan to make the check more comprehensive after we get the initial set of test done.

File: testing/citest/spinnaker_testing/azure_scenario_support.py (hunk 6 above)
Comment (ewiseblatt) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: Maybe default the value to an environment variable as an option?
If you pass keys on the command-line then they become visible in all the places that record command lines (process status, shell history, etc). You can still learn the key in your jenkins task doing the scp, but set an environment variable rather than passing another flag.

File: testing/citest/spinnaker_testing/spinnaker_test_scenario.py [hunk 12]
```diff
@@ -35,12 +35,12 @@
...
-    appengine_scenario_support.AppEngineScenarioSupport,
-    kubernetes_scenario_support.KubernetesScenarioSupport,
-    openstack_scenario_support.OpenStackScenarioSupport,
+    # aws_scenario_support.AwsScenarioSupport,
+    # # appengine depends on google so order it after
+    # google_scenario_support.GoogleScenarioSupport,
+    # appengine_scenario_support.AppEngineScenarioSupport,
+    # kubernetes_scenario_support.KubernetesScenarioSupport,
+    # openstack_scenario_support.OpenStackScenarioSupport,
     azure_scenario_support.AzureScenarioSupport
```
Comment (ewiseblatt) [inline testing/citest/spinnaker_testing/spinnaker_test_scenario.py]: are these a problem>
Why did you need to comment them out? Is there something I need to fix/change?
(and, obviously uncomment them)

**Comment by rguthriemsft** [comment]
@EricJizbaMSFT @ewiseblatt PTAL, all test are working now.

File: testing/citest/spinnaker_testing/azure_scenario_support.py (hunk 6 above)
Comment (ewiseblatt) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: I still think you should default the app key to an environment variable.
Passing the other parameters on the commandline is fine, but you wouldnt want to handle the credentials the same way so it just creates more work. You have a scheme where you get them all the same way from the deployment, but that wont work in general (e.g. when the deployment isnt azure, which wont always be the case). It isnt extra work because you dont have to use it, in which case you are back to where you already are here.

File: testing/citest/spinnaker_testing/azure_scenario_support.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        help='The service principal secret.')
+    builder.add_argument(
+        '--test_azure_rg_location', 
+        default = defaults.get('TEST_AZURE_RG_LOCATION', 'westus'),
+        help='The location of the azure resource group where test resources should be created.')
+    builder.add_argument(
+        '--azure_storage_account_name', 
+        dest='azure_storage_account_name',
+        help='The name of the Azure storage account used by Front50 in Spinnaker.')
+    builder.add_argument(
```
Comment (ewiseblatt) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: ditto for environment variable default value here too if this is a sensitive value.

File: testing/citest/spinnaker_testing/spinnaker.py [hunk 14]
```diff
@@ -371,7 +371,7 @@ def new_native_instance(cls, name, status_factory, base_url):
       return None
 
     logger.info('%s is available at %s', name, base_url)
-    env_url = os.path.join(base_url, 'env')
+    env_url = os.path.join(base_url, 'resolvedEnv')
```
Comment (ewiseblatt) [inline testing/citest/spinnaker_testing/spinnaker.py]: I made this fix this several days ago. Sync your repo.

File: testing/citest/tests/azure_smoke_test.py [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+            description= 'Test - create application {app}'.format(app=self.bindings['TEST_APP']),
+            application=self.bindings['TEST_APP'])
+
+    builder = az.AzContractBuilder(self.az_observer)
+    (builder.new_clause_builder(
+        'Application Created', retryable_for_secs=30)
+      .collect_resources(
+          az_resource='storage',
+          command='blob',
+          args=['exists', '--container-name', 'front50',
```
Comment (ewiseblatt) [inline testing/citest/tests/azure_smoke_test.py]: What is "front50" here? Is this the name of an azure resource (e.g. are you assuming spinnaker is configured with a container named "front50" ? Is this something you hardcode into spinnaker, or was this a deployment choice? If a deployment choice, it should be parameterized (e.g. get it from the agent.deployed_config) so you can run the test against any deployment without having to overprescribe configuration.

File: testing/citest/tests/azure_smoke_test.py [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+    (builder.new_clause_builder(
+        'Security Group Created', retryable_for_secs=30)
+    .collect_resources(
+         az_resource='network',
+         command='nsg',
+         args=['show', '--name', self.TEST_SECURITY_GROUP, '--resource-group', self.TEST_SECURITY_GROUP_RG],
+         no_resources_ok=True)
+    #sec grp name matches expected
+    .contains_match({'name': jp.STR_SUBSTR(self.TEST_SECURITY_GROUP)})
+    .contains_path_pred(
```
Comment (ewiseblatt) [inline testing/citest/tests/azure_smoke_test.py]: This looks older than our conversation earlier today.
I'm going to stop here and ping you because I think this is out of date.

Comment (ewiseblatt) [inline testing/citest/tests/azure_smoke_test.py]: My bad, my browser was clipping #207 and I didnt realize it. This looks current.

File: testing/citest/tests/azure_smoke_test.py [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+    """
+    super(AzureSmokeTestScenario, self).__init__(bindings, agent)
+    bindings = self.bindings
+
+    # create default sec grp name  
+    # Spinnaker-azure appends region to actual RG used by an app to support multi-region
+    # pylint: disable=invalid-name
+    self.TEST_SECURITY_GROUP = 'sec_grp_' + self.bindings['TEST_APP']
+    self.TEST_SECURITY_GROUP_RULE_1=self.TEST_SECURITY_GROUP + '_RULE_1'
+    self.TEST_SECURITY_GROUP_RG = self.bindings['TEST_APP'] + '-' + self.bindings['TEST_AZURE_RG_LOCATION']
```
Comment (ewiseblatt) [inline testing/citest/tests/azure_smoke_test.py]: I assume the  GROUP_RULE_1 is an arbitrary name you are picking for this test
and he GROUP_RG is a specific name generated by your clouddriver heuristics.

If this is the case, I suggest you add a method to spinnaker_testing.azure_support that knows this policy and can return the names. it will serve as some documentation, add clarity to future tests, and help distinguish the arbitrary names from the intentional ones.

If you are following standard spinnaker frigga naming and I'm missing that mapping then you can use spinnaker_testing.frigga.Naming to generate the names.

File: testing/citest/tests/azure_smoke_test.py (hunk 15 above)
Comment (rguthriemsft) [inline testing/citest/tests/azure_smoke_test.py]: its the name of the container that is set in spinnaker-local.yml.  we default it to front50 so its easy to understand what the storage container is for in the storage account.

File: testing/citest/spinnaker_testing/azure_scenario_support.py (hunk 13 above)
Comment (rguthriemsft) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: I ave removed the subscription parameters.  I will look into the storage account key and putting into environment variable.  Can we push this and that is something I will look into immediately as I am unsure if we consider the storage account key sensitive info

File: testing/citest/spinnaker_testing/azure_scenario_support.py (hunk 6 above)
Comment (rguthriemsft) [inline testing/citest/spinnaker_testing/azure_scenario_support.py]: app key is removed, we don't need it in the test as we assume you have already logged into the cli before running the test.
